### PR TITLE
meraki_network - Parameter change for combined network type

### LINKED ISSUE #49149: meraki_network - Update combined network type
##### SUMMARY

Combined network type automatically sets type to `switch wireless appliance` however not all combined templates are set up this way.  It appears that Combined only supports devices that are associated to an Organization.  Our org does not have wireless and the network would not bind to the configuration template until I removed wireless from the type payload.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

meraki_network

##### ANSIBLE VERSION

```paste below
ansible 2.7.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/ansible/plugins/modules/ntc-ansible', u'/usr/local/lib/python2.7/dist-packages/ara/plugins/modules', u'/home/ansible/.ansible/roles/mkouhei.include_csv']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15rc1 (default, Nov 12 2018, 14:31:15) [GCC 7.3.0]
```

##### CONFIGURATION

```paste below
ansible-config dump --only-changed
DEFAULT_ACTION_PLUGIN_PATH(/etc/ansible/ansible.cfg) = [u'/usr/local/lib/python2.7/dist-packages/ara/plugins/actions']
DEFAULT_CALLBACK_PLUGIN_PATH(/etc/ansible/ansible.cfg) = [u'/usr/local/lib/python2.7/dist-packages/ara/plugins/callbacks']
DEFAULT_CALLBACK_WHITELIST(/etc/ansible/ansible.cfg) = [u'profile_tasks', u'dense', u'selective']
DEFAULT_FORKS(/etc/ansible/ansible.cfg) = 10
DEFAULT_LOCAL_TMP(/etc/ansible/ansible.cfg) = /tmp/.ansible/tmp/ansible-local-14602D6jlLw
DEFAULT_MODULE_PATH(/etc/ansible/ansible.cfg) = [u'/usr/share/ansible/plugins/modules/ntc-ansible', u'/usr/local/lib/python2.7/dist-packages/ara/plugins/modules', u'/home/ansible/.ansible/roles/mkouhei.include_csv']
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
RETRY_FILES_ENABLED(/etc/ansible/ansible.cfg) = False
```

##### STEPS TO REPRODUCE

Create a new network with type combined on an organization that only has 2 of the 3 types
Attempt to bind the network to a configuration template

```yaml
---

- name: GET DATE AND GATE
  connection: local
  hosts: localhost

  vars:
    my_key: 'abc12345'
    my_org: 'Meraki Organization'
    my_net_name: 'Meraki_Network'
    address: "1060 W. Addison St., Chicago, IL"
    mx_serial: "ABC-123"
    mx_name: "ABC-RT01"
    ms_name: "ABC-SW01"
    ms_serial: "ABC-456"
    config_template: "Standard_Site"

  tasks:

    - name: "Create meraki network {{ my_net_name }}"
      meraki_network:
        auth_key: " {{ my_key }}"
        state: present
        org_name: "{{ my_org }}"
        net_name: "{{ my_net_name }}"
        type: combined
        timezone: America/Chicago
        tags: production, central
      delegate_to: localhost

    - name: Pause for Meraki API rate limiting
      pause:
        seconds: 1

    - name: "Bind {{ mx_serial }} to {{ my_net_name }}"
      meraki_device:
        auth_key: "{{ my_key }}"
        org_name: "{{ my_org }}"
        net_name: "{{ my_net_name }}"
        serial: "{{ mx_serial }}"
        state: present
      delegate_to: localhost

    - name: Pause for Meraki API rate limiting
      pause:
        seconds: 1

    - name: "Bind {{ ms_serial }} to {{ my_net_name }}"
      meraki_device:
        auth_key: "{{ my_key }}"
        org_name: "{{ my_org }}"
        net_name: "{{ my_net_name }}"
        serial: "{{ ms_serial }}"
        state: present
      delegate_to: localhost

    - name: Pause for Meraki API rate limiting
      pause:
        seconds: 1

    - name: "Bind network {{ my_net_name }} to configuration template {{ config_template }}"
      meraki_config_template:
        auth_key: "{{ my_key }}"
        state: present
        org_name: "{{ my_org }}"
        net_name: "{{ my_net_name }}"
        config_template: "{{ config_template }}"
        auto_bind: yes
      delegate_to: localhost
```

##### EXPECTED RESULTS

Network to bind to config template

##### ACTUAL RESULTS

Error received due to incompatible network types

```paste below
fatal: [localhost -> localhost]: FAILED! => {
    "body": {
        "errors": [
            "'Meraki_Network' cannot be bound to 'Standard_Site' because they have incompatible network types."
        ]
    }, 
    "changed": true, 
    "invocation": {
        "module_args": {
            "auth_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "auto_bind": true, 
            "config_template": "Standard_Site", 
            "follow_redirects": "all", 
            "host": "api.meraki.com", 
            "net_id": null, 
            "net_name": "Meraki_Network", 
            "org_id": null, 
            "org_name": "Meraki Organization", 
            "output_level": "normal", 
            "protocol": "https", 
            "state": "present", 
            "timeout": 30, 
            "use_https": true, 
            "use_proxy": false, 
            "validate_certs": true
        }
    }, 
    "msg": "Request failed for https://api.meraki.com/api/v0/networks/<snip>/bind: 400 - HTTP Error 400: Bad Request", 
    "response": "HTTP Error 400: Bad Request", 
    "status": 400
}

```

### PR DESCRIPTION
**Pull request by kbreit**
##### SUMMARY
`meraki_network` used to have a network type of combined, but that hard coded the combined types. This pull request adds a new parameter, `types`, which allows the playbook author to apply a list of network types for greater flexibility

Fixes #49149
@ericdost - If you can, please test using my branch. Another option I have is to keep the `type` parameter as is and allow the playbook developer to specify the space delimited types. Let me know if you have a preference one way or another.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
meraki_network

### CODE REVIEW

**Comment by kbreit** [comment]
@dagwieers The test timed out. Is that due to bug in my code or error on Shippable's side?

**Comment by kbreit** [comment]
I have some code written to allow the idempotency check to run. However, I have a few open PRs which will affect this code so I'd rather wait until they are merged and test using the latest and greatest.

Waiting for the following merges:
https://github.com/ansible/ansible/pull/48971
https://github.com/ansible/ansible/pull/48820
https://github.com/ansible/ansible/pull/48394
https://github.com/ansible/ansible/pull/48391 (merged)

**Comment by kbreit** [comment]
I would like to get this merged, despite waiting for the other PR’s to be merged. I can deal with any conflicts in the other PR’s.

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (dagwieers) [inline lib/ansible/modules/network/meraki/meraki_network.py]: We tend to leave spaces around list content in YAML ofr readability, but it's debatable as PEP8 does not allow this in Python.
```suggestion
        choices: [ appliance, switch, wireless ]
```

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (dagwieers) [inline lib/ansible/modules/network/meraki/meraki_network.py:111]: This really should be defined as a list.
And that's good news, because it supports comma-separated string values automatically !

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (dagwieers) [inline lib/ansible/modules/network/meraki/meraki_network.py]: So this ought to be a list.

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (dagwieers) [inline lib/ansible/modules/network/meraki/meraki_network.py]: We tend to sort items in lists alphabetically if the order is of no importance.

File: test/integration/targets/meraki_network/tasks/main.yml
Comment (dagwieers) [inline test/integration/targets/meraki_network/tasks/main.yml]:
```suggestion
      org_name: '{{ test_org_name }}'
```

**Review by dagwieers — CHANGES_REQUESTED** [review]
Looking good, small remarks.

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_network.py:111]: So are you good with how it is?

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_network.py]: I'm trying to work this out and can't figure how to do this without breaking backward compatibility. If I change `type` to a list it won't accept the strings as I've had it.

Comment (dagwieers) [reply lib/ansible/modules/network/meraki/meraki_network.py]: Yes it will, it supports comma-separated values out of the box.

Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_network.py]: Got it now.

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (dagwieers) [inline lib/ansible/modules/network/meraki/meraki_network.py:4]: So we tend not to do these kinds of updates.

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (dagwieers) [inline lib/ansible/modules/network/meraki/meraki_network.py:195]: This is not backward compatible. Fine if you want to continue regardless :-)

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (dagwieers) [inline lib/ansible/modules/network/meraki/meraki_network.py]: Why is tags not defined as a list ?

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_network.py:195]: If I'm going to do it, now is probably the time before many people are using these modules.

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_network.py]: I've decided I will update this to be a list. Again, break compatibility now.

**Comment by kbreit** [comment]
@dagwieers PR is updated. Please review.

File: lib/ansible/modules/network/meraki/meraki_network.py
Comment (dagwieers) [reply lib/ansible/modules/network/meraki/meraki_network.py]: This does not break compatibility. Moving from a list to a string does break compatibility, but not this.

**Pull request merged**
